src/nodes.coffee

```diff
@@ -1611,70 +1611,133 @@ exports.Code = class Code extends Base
-
-    # Handle bound functions early.
-    if @bound and not @context
-      @context = '_this'
-      wrapper = new Code [new Param new IdentifierLiteral @context], new Block [this]
-      boundfunc = new Call(wrapper, [new ThisLiteral])
-      boundfunc.updateLocationDataIfMissing @locationData
-      return boundfunc.compileNode(o)
+    if @bound
+      @context = o.scope.method.context if o.scope.method?.bound
+      @context = 'this' unless @context
 
     o.scope         = del(o, 'classScope') or @makeScope o.scope
     o.scope.shared  = del(o, 'sharedScope')
     o.indent        += TAB
     delete o.bare
     delete o.isExistentialEquals
-    params = []
-    exprs  = []
-    for param in @params when param not instanceof Expansion
-      o.scope.parameter param.asReference o
-    for param in @params when param.splat or param instanceof Expansion
-      for p in @params when p not instanceof Expansion and p.name.value
-        o.scope.add p.name.value, 'var', yes
-      splats = new Assign new Value(new Arr(p.asReference o for p in @params)),
-                          new Value new IdentifierLiteral 'arguments'
-      break
-    for param in @params
-      if param.isComplex()
-        val = ref = param.asReference o
-        val = new Op '?', ref, param.value if param.value
-        exprs.push new Assign new Value(param.name), val, '=', param: yes
+    params           = []
+    exprs            = []
+    paramsAfterSplat = []
+    haveSplatParam   = no
+
+    # Check for duplicate parameters.
+    paramNames = []
+    @eachParamName (name, node) =>
+      node.error "multiple parameters named '#{name}'" if name in paramNames
+      paramNames.push name
+
+    # Parse the parameters, adding them to the list of parameters to put in the
+    # function definition; and dealing with splats or expansions, including
+    # adding expressions to the function body to declare all parameter
+    # variables that would have been after the splat/expansion parameter.
+    for param, i in @params
+      # Was `...` used with this parameter? (Only one such parameter is allowed
+      # per function.) Splat/expansion parameters cannot have default values,
+      # so we need not worry about that.
+      if param.splat or param instanceof Expansion
+        if haveSplatParam
+          param.error 'only one splat or expansion parameter is allowed per function definition'
+        else if param instanceof Expansion and @params.length is 1
+          param.error 'an expansion parameter cannot be the only parameter in a function definition'
+
+        haveSplatParam = yes
+        if param.splat
+          params.push ref = param.asReference o
+          splatParamName = fragmentsToText ref.compileNode o
+          if param.isComplex() # Parameter is destructured or attached to `this`
+            exprs.push new Assign new Value(param.name), ref, '=', param: yes
+            # TODO: output destrucutred parameters as is, *unless* they contain
```

Comment:
I was messing about yesterday with compiling destructuring to ES2015 in all cases, and discovered that `this` parameters can be compiled quite neatly when destructuring:

```coffee
{ @a } = b
[ @a ] = b
```

This can become:

```js
{ a: this.a } = b
[ this.a ] = b
```